rust/processor/src/db_writer.rs

```diff
@@ -0,0 +1,309 @@
+        res
+    }
+}
+
+pub fn diesel_error_to_metric_str(error: &Error) -> &'static str {
+    match error {
+        Error::DatabaseError(kind, _) => match kind {
+            DatabaseErrorKind::UniqueViolation => "DatabaseError::UniqueViolation",
+            DatabaseErrorKind::ForeignKeyViolation => "DatabaseError::ForeignKeyViolation",
+            DatabaseErrorKind::UnableToSendCommand => "DatabaseError::UnableToSendCommand",
+            DatabaseErrorKind::SerializationFailure => "DatabaseError::SerializationFailure",
+            DatabaseErrorKind::ReadOnlyTransaction => "DatabaseError::ReadOnlyTransaction",
+            DatabaseErrorKind::NotNullViolation => "DatabaseError::NotNullViolation",
+            DatabaseErrorKind::CheckViolation => "DatabaseError::CheckViolation",
+            DatabaseErrorKind::ClosedConnection => "DatabaseError::ClosedConnection",
+            DatabaseErrorKind::Unknown => "DatabaseError::Unknown",
+            _ => "DatabaseError::Other",
+        },
+        Error::NotFound => "NotFound",
+        Error::QueryBuilderError(_) => "QueryBuilderError",
+        Error::DeserializationError(_) => "DeserializationError",
+        Error::SerializationError(_) => "SerializationError",
+        Error::InvalidCString(_) => "InvalidCString",
+        Error::RollbackTransaction => "RollbackTransaction",
+        Error::AlreadyInTransaction => "AlreadyInTransaction",
+        Error::RollbackErrorOnCommit { .. } => "RollbackErrorOnCommit",
+        Error::NotInTransaction => "NotInTransaction",
+        Error::BrokenTransactionManager => "BrokenTransactionManager",
+        _ => "Other",
+    }
+}
+
+// A holder struct for processors db writing so we don't need to keep adding new params
+#[derive(Clone)]
+pub struct DbWriter {
+    pub processor_name: &'static str,
+    pub db_pool: PgDbPool,
+    pub query_sender: AsyncSender<QueryGenerator>,
+    pub per_table_chunk_sizes: AHashMap<String, usize>,
+    pub skip_tables: AHashSet<String>,
+}
+
+impl DbWriter {
+    pub fn new(
+        processor_name: &'static str,
+        db_pool: PgDbPool,
+        query_sender: AsyncSender<QueryGenerator>,
+        per_table_chunk_sizes: AHashMap<String, usize>,
+        skip_tables: AHashSet<String>,
+    ) -> Self {
+        Self {
+            processor_name,
+            db_pool,
+            query_sender,
+            per_table_chunk_sizes,
+            skip_tables,
+        }
+    }
+
+    pub fn chunk_size<Item: field_count::FieldCount>(&self, table_name: &str) -> usize {
+        get_config_table_chunk_size::<Item>(table_name, &self.per_table_chunk_sizes)
+    }
+
+    pub async fn send_in_chunks<Item>(&self, table_name: &'static str, items_to_insert: Vec<Item>)
+    where
+        Item: field_count::FieldCount
+            + serde::Serialize
+            + for<'de> serde::Deserialize<'de>
+            + Send
+            + Sync
+            + Clone
+            + 'static,
+        Vec<Item>: DbExecutable,
```

Comment:
how does contraints on a Vec<T> work? i thought you could only constraint generics.

Reply:
you can `impl` on a `Vec<Item>` :-)